core/chaincode/lifecycle/lifecycle.go

```diff
@@ -675,6 +675,9 @@ func (ef *ExternalFunctions) InstallChaincode(chaincodeInstallPackage []byte) (*
 	}
 
 	buildStatus, ok := ef.BuildRegistry.BuildStatus(packageID)
+	if buildStatus != nil && (buildStatus.Err() != nil) {
```

Comment:
This is racey.  What if the chaincode is currently being built, and fails?  Until the done channel closes, we don't know whether the result of buildStatus.Err() is error free or simply uninitialized.

We need to wait until after the build has completed, (line 682/685)

Then, if we were not the ones who initiated the build (ie `ok`) and the build failed (ie `Err() != nil`).  Then we should reset and rebuild.

Reply:
This still seems racey to me.  If the buildStatus is non-nil, it could be that the build is still ongoing, and may yet fail, so it is not safe to read from Err() yet.  We must wait until the done channel closes.

Reply:
Doesn't the build lock by package id prevent that? We make sure that only one chaincode per package id can be built at a given time. The first caller may have a build running but they hold the buildLock. Once the buildLock is unlocked by the first caller, the second caller grabs the lock, gets the build status, and continues from there.

Reply:
I see what you're saying.  Since we are in the new lifecycle, there should not be other competing threads trying to race to build the chaincode, other than ourselves.

Still, I think this isn't quite right.  `buildStatus` is never nil.  So checking for non-nil is redundant.  I guess I would do:

```
if !ok {
    // Another invocation of lifecycle has concurrently installed this same package-id
   <-buildStatus.Done() // this should always already be closed
   if buildStatus.Err() == nil {
       return nil, errors.Errorf("chaincode already successfully installed")
  }
  buildStatus, _ = ef.BuildRegistry.ResetBuildStatus(packageID)
}
```